### ISSUE TITLE
DeCo 配置下 beam search 解码结果异常，greedy / nucleus 正常，是否参数设置问题

### ISSUE THREAD

**Issue by HanKezh** [zjunlp/Deco#10]
---

您好，我在使用 LLaVA-v1.5 模型并集成 DeCo 解码策略进行实验时，发现如下问题：

在相同主干参数设置下，**greedy** 与 **nucleus sampling** 解码的结果表现正常；但 **beam search**（`num_beams=3`）的评估结果异常偏高，表现明显不符合预期，特此反馈并寻求建议。

---

### beam Search 配置与结果（表现异常）

```bash
python chair_llava5.py \
  --model-path <路径>/llava-v1.5-7b \
  --answers-file ./output/llava_beam_deco.jsonl \
  --log_path ./logs/llava_beam_deco \
  --conv-mode llava_v1 \
  --temperature 1.0 \
  --top_p 0.9 \
  --top_k 20 \
  --num_beams 3 \
  --max_new_tokens 512 \
  --alpha 0.5 \
  --threshold_top_p 0.9 \
  --threshold_top_k 20 \
  --start_layer 20 \
  --end_layer 29
```

**评估结果（CHAIR）：**
```
CHAIRs    : 50.4
CHAIRi    : 13.1
Recall    : 80.1
Len       : 97.4
```

---

### Greedy 配置与结果（表现正常）

```bash
python chair_llava7.py \
  --model-path <路径>/llava-v1.5-7b \
  --answers-file ./output/llava_greedy_deco.jsonl \
  --log_path ./logs/llava_greedy_deco \
  --conv-mode llava_v1 \
  --temperature -1 \
  --top_p 0.9 \
  --top_k 20 \
  --num_beams 1 \
  --max_new_tokens 512 \
  --alpha 0.5 \
  --threshold_top_p 0.9 \
  --threshold_top_k 20 \
  --start_layer 20 \
  --end_layer 29
```

**评估结果：**
```
CHAIRs    : 35.8
CHAIRi    : 10.0
Recall    : 74.2
Len       : 96.8
```

---

### Nucleus Sampling 配置与结果（表现正常）

```bash
python chair_llava2.py \
  --model-path <路径>/llava-v1.5-7b \
  --answers-file ./output/llava_nucleus_deco.jsonl \
  --log_path ./logs/llava_nucleus_deco \
  --conv-mode llava_v1 \
  --temperature 1.0 \
  --top_p 0.9 \
  --top_k 20 \
  --num_beams 1 \
  --max_new_tokens 512 \
  --alpha 0.5 \
  --threshold_top_p 0.9 \
  --threshold_top_k 20 \
  --start_layer 20 \
  --end_layer 29
```

**评估结果：**
```
CHAIRs    : 43.4
CHAIRi    : 12.4
Recall    : 73.7
Len       : 98.6
```

---

### 疑问：

从理论上看，beam search 应具备更稳定、多样的生成能力，但目前结果反而不如 greedy 和 nucleus。是否可能由以下原因导致：

- `temperature=1.0` 是否与 beam 策略存在冲突？
- `alpha`、`threshold_top_p/k` 是否影响 beam 路径中的 token 筛选或 logits 融合效果？

希望能确认是否为参数设置问题，或是否推荐特定超参组合以改善 beam 表现。非常感谢您对 DeCo 项目的持续优化！

---

**Comment by HanKezh** [comment]
以及这是我的chair_llava.py


```bash
import argparse
import torch
import os
import json
from tqdm import tqdm
import shortuuid
import sys
import os

sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
# print(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from llava.constants import (
    IMAGE_TOKEN_INDEX, # IMAGE_TOKEN_INDEX = -200
    DEFAULT_IMAGE_TOKEN, # DEFAULT_IMAGE_TOKEN = "<image>"
    DEFAULT_IM_START_TOKEN, # DEFAULT_IM_START_TOKEN = "<im_start>"
    DEFAULT_IM_END_TOKEN, # DEFAULT_IM_END_TOKEN = "<im_end>"
    IMAGE_PLACEHOLDER, # IMAGE_PLACEHOLDER = "<image-placeholder>"
)
from llava.conversation import conv_templates, SeparatorStyle
from llava.model.builder import load_pretrained_model
from llava.utils import disable_torch_init
from llava.mm_utils import tokenizer_image_token, process_images, get_model_name_from_path, KeywordsStoppingCriteria

from PIL import Image
import base64
import requests

from io import BytesIO
import re
import math
import torch.distributed as dist
from utils import dist_util
from utils.logger import create_logger
from glob import glob
from transformers import set_seed

device = "cuda:1"
def image_parser(args):
    out = args.image_file.split(args.sep)
    return out


def load_image(image_file):
    # 预处理
    if image_file.startswith("http") or image_file.startswith("https"):
        response = requests.get(image_file)
        image = Image.open(BytesIO(response.content)).convert("RGB")
    else:
        image = Image.open(image_file).convert("RGB")
    return image


def load_images(image_files):
    # 图片列表
    out = []
    for image_file in image_files:
        image = load_image(image_file)
        out.append(image)
    return out





def eval_model(args):
    
    # set up gpu and logging
    
    dist_util.setup_dist(args)
    device = dist_util.device()

    # Setup an experiment folder:
    if dist.get_rank() == 0:
        os.makedirs(
            args.log_path, exist_ok=True
        )  # Make results folder (holds all experiment subfolders)
        model_string_name = args.model_path.split("/")[-1]
        experiment_index = len(glob(f"{args.log_path}/{model_string_name}/*"))
        experiment_dir = f"{args.log_path}"  # Create an experiment folder
        os.makedirs(experiment_dir, exist_ok=True)
        logger = create_logger(experiment_dir)
        logger.info(f"Experiment directory created at {experiment_dir}")
    else:
        logger = create_logger(None)
    
    # Model
    disable_torch_init()
    model_path = os.path.expanduser(args.model_path)
    model_name = get_model_name_from_path(model_path)
    tokenizer, model, image_processor, context_len = load_pretrained_model(model_path, args.model_base, model_name,device="cuda:1")
    answers_file = os.path.expanduser(args.answers_file)
    os.makedirs(os.path.dirname(answers_file), exist_ok=True)    
    ans_file = open(answers_file, "w", encoding="utf-8")
    with open("/home/user/MLLMs/Deco/opera_log/llava-1.5/ours.jsonl", "r", encoding="utf-8") as f:
        for _, data_line in tqdm(enumerate(f.readlines()),total=500):
            line = json.loads(data_line)
            idx = line["image_id"]
            # image_file = ".../val2014/COCO_val2014_" + str(idx).zfill(12) + ".jpg"
            image_file = f"/home/userMLLMs/data/val2014/COCO_val2014_{str(idx).zfill(12)}.jpg"
            qs = "Please describe this image in detail."#line["query"]
            cur_prompt = qs
    
            if model.config.mm_use_im_start_end:
                qs = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + '\n' + qs
            else:
                qs = DEFAULT_IMAGE_TOKEN + '\n' + qs

            conv = conv_templates[args.conv_mode].copy()
            conv.append_message(conv.roles[0], qs)
            conv.append_message(conv.roles[1], None)
            prompt = conv.get_prompt()

            input_ids = tokenizer_image_token(prompt, tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt').unsqueeze(0).cuda(1)
            image = Image.open(image_file)
            image_tensor = image_processor.preprocess(image, return_tensors='pt')['pixel_values'][0]            
            stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
            keywords = [stop_str]
            stopping_criteria = KeywordsStoppingCriteria(keywords, tokenizer, input_ids)

            with torch.inference_mode():
                with torch.no_grad():
                    output_dict = model.generate(
                        input_ids,
                        images=image_tensor.unsqueeze(0).half().cuda(5),
                        do_sample=True if args.temperature > 0 else False,
                        temperature=args.temperature,
                        top_p=args.top_p,
                        num_beams=args.num_beams,
                        max_new_tokens=args.max_new_tokens,
                        return_dict_in_generate=True,
                        output_hidden_states=True,
                        stopping_criteria=[stopping_criteria],
                        use_deco = True,
                        alpha = args.alpha,
                        threshold_top_p=args.threshold_top_p, 
                        threshold_top_k=args.threshold_top_k,
                        early_exit_layers=[i for i in range(args.start_layer, args.end_layer)],
                        return_dict=True
                        )
                
            output_ids = output_dict.sequences
            input_token_len = input_ids.shape[1]
            outputs = tokenizer.batch_decode(
                    output_ids[:, input_token_len:], skip_special_tokens=True
                )[0]
            outputs = outputs.strip()
                
                
            logger.info(f"[{image_file}]")
            logger.info(f"prompt: {cur_prompt}") 
            logger.info(f"text: {outputs}")  
            # res_dict = {"id": idx,"response": outputs}
            res_dict = {"image_id": idx, "caption": outputs}

            # ans_file.write(json.dumps(res_dict, ensure_ascii=False) + "\n")
            ans_file.write(json.dumps(res_dict, ensure_ascii=False) + "\n")
            ans_file.flush()  # 可选，确保每条都即时写入
        ans_file.close()  # 最后关闭文件

                


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--model-path", type=str, default=".../llava-v1.5-7b")
    parser.add_argument("--model-base", type=str, default=None)
    parser.add_argument("--answers-file", type=str, default="")
    parser.add_argument("--conv-mode", type=str, default="llava_v1")
    parser.add_argument("--num-chunks", type=int, default=1)
    parser.add_argument("--chunk-idx", type=int, default=0)
    parser.add_argument("--temperature", type=float, default=-1) # 可以考虑调大温度试试
    parser.add_argument("--top_p", type=float, default=None)
    parser.add_argument("--top_k", type=int, default=None)
    parser.add_argument("--num_beams", type=int, default=1)
    parser.add_argument("--max_new_tokens", type=int, default=512)
    parser.add_argument("--log_path", type=str, default="...")
    parser.add_argument("--batch_size", type=int, default=1)
    parser.add_argument("--alpha", type=float, default=0.6)
    parser.add_argument("--threshold_top_p", type=float, default=0.9)
    parser.add_argument("--threshold_top_k", type=int, default=20)
    parser.add_argument("--start_layer", type=int, default=20)
    parser.add_argument("--end_layer", type=int, default=29)
    parser.add_argument("--seed", type=int, default=42)

    args = parser.parse_args()
    set_seed(args.seed)
    eval_model(args)
```

**Comment by sunnychenxiwang** [comment]
据我所知，beam search会根据temperature的值（如果大于0则是beamsample，如果不设置或者use_sample设置为False才是beamsearch）采用两种策略。具体的你可以参考transformers/generation/utils.py文件中的，解码策略采用时的逻辑分支。

**Comment by HanKezh** [comment]
非常感谢您的回复 我再次跑一下实验！

**Comment by HanKezh** [comment]
是我的温度和beam数出了问题，我新跑的一次是num_beams=5 temp=-1，最后结果和作者的结果基本完全一致，是非常好的一篇工作 ！！！祝福作者大大paper多多
